### DISCUSSION TITLE
Using Tectonic with VS Code

### DISCUSSION THREAD

**Discussion by ujjwal123123** [tectonic-typesetting/tectonic#896] [Q&A] [answered]
I have been using Tectonic from terminal itself, manually entering compilation commands and editing with files with gedit. But since VS Code has great extension like [TexLab](https://github.com/latex-lsp/texlab-vscode) and [LaTeX Workshop](https://github.com/James-Yu/LaTeX-Workshop), I wanted to try it and could not get those extensions to work with Tectonic.

It would be helpful if someone could share instructions for configuring one of these extensions to work with Tectonic.

**Comment by jarmos-san** [comment, marked as answer]
**EDIT**: So, after taking a look at [this answer](https://github.com/tectonic-typesetting/tectonic/discussions/896#discussioncomment-2847604) I wanted to update my workflow to adhere to `tectonic`'s v2 interface as well. This answer is an attempt to do so & if you would like to check out the older v1 way of doing things, see [this answer (comment)](https://github.com/tectonic-typesetting/tectonic/discussions/896#discussioncomment-2850685).

## Creating a Base to Start Working From

You should generate a base work environment by invoking the `tectonic -X new` command. It should create the follow directory structure:

```console
.
├── src
│   ├── _preamble.tex
│   ├── _postamble.tex
│   └── index.tex
└── Tectonic.toml
```
You'll figure out the significance of the `.tex` files eventually. But it's important to pay special heed to the `Tectonic.toml` file since this is what'll help `tectonic` with the file names! The file names are also important without which [Latex-Workshop](https://github.com/James-Yu/LaTeX-Workshop) willn't function properly.

That said, here's a bare minimum `Tectonic.toml` file:
```toml
[doc]
name = "example-book"
# The following bundle will be generated automatically by the "tectonic -X new" command. DO NOT change it!
bundle = "https://some-remote-url.tar"

[[output]]
# DO NOT change! The VSCode plugin mightn't work as expected if you do so.
name = "index"
type = "pdf"
```
You can update & modify the file as per your needs by referring to the [official documentations](https://tectonic-typesetting.github.io/book/latest/ref/tectonic-toml.html) on the `Tectonic.toml` file.

**NOTE**: It's IMPORTANT you don't change the value of `output.name` unless you know what you're doing. Else Latex-Workshop won't open the in-built PDF viewer properly, more on it in a later section.

## Update the Contents of the Generated `.tex` Files

The `.tex` files `tectonic` generates is a good starting point for us to start with. You can read up more on the `tectonic -X new` command in it's [official documentations](https://tectonic-typesetting.github.io/book/latest/v2cli/new.html). The documentations speaks of a "[suggested framework](https://tectonic-typesetting.github.io/book/latest/ref/documents.html)".

Following the framework means, you needn't maintain an individual `preamble.tex` file & then `\include` them in a `main.tex` file (see _"[Module Documents](https://en.wikibooks.org/wiki/LaTeX/Modular_Documents)")._ `\include` is known to break workflows & is prone to errors which are difficult to fix especially for modular project structures.

But fortunately, `tectonic` takes care of those issues! Hence, all you need is a single command - `tectonic -X build` to compile your PDF file.

## Configuring VSCode to Use `tectonic`

To use Latex in VSCode, I personally suggest using [Latex-Workshop](https://github.com/James-Yu/LaTeX-Workshop) because of how feature-full it is. The alternative is [TexLabs](https://github.com/latex-lsp/texlab-vscode) which isn't as feature-full but provides better LSP than Latex-Workshop.

That said, following are the VSCode configurations you need to copy-paste into your [`settings.json`](https://code.visualstudio.com/docs/getstarted/settings) file:

```json
  "latex-workshop.latex.recipe.default": "tectonic",
  "latex-workshop.latex.autoBuild.run": "onSave",
  "latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/index",
  "latex-workshop.view.pdf.viewer": "tab",
  "latex-workshop.latex.recipes": [
    {
      "name": "tectonic",
      "tools": ["tectonic"]
    }
  ],
  "latex-workshop.latex.tools": [
    {
      "name": "tectonic",
      "command": "tectonic",
      "args": ["-X", "build", "--keep-intermediates", "--keep-logs"],
      "env": {}
    }
  ]
```
The `"latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/index"` settings requires special attention here. Remember when I mentioned not to change the value of `output.name` in the `Tectonic.toml` file? It;s because of this VSCode setting which enables the plugin to search for the compiled `.pdf` file in the right place.

That said, at the end of the day when you're draft is complete, Press the compile button in VSCode & find the `index.pdf` file under the `build/index` directory. Rename the file to something more apt & you're good to go!

The updated answer turned out to be lengthier than I expected but I hope it helps you guys out there to use VSCode & LaTex for writing your PhD thesis! :sweat_smile:

**CREDITS**: Thanks to @LeCyberDucky for sharing his configurations which makes use of the v2 interface of `tectonic`. I updated this answer (and my current workflow) by taking inspirations from his config.

**Reply by roaldarbol** [reply]
I'm trying to get this to work, but keep getting `error: *** (job aborted, no legal \end found)`. I've started with a complete blank example using `tectonic -X new`, and I can get it compiled with `tectonic` in the command line. It seems it's not registering that there's multiple files that need to be checked, and tries to compile using `_preamble.tex` as the main file. There might be something obvious that I'm missing, but I simply can't figure out what it is. Have you seen this before?

<details><summary>Log</summary>

```zsh 
**
(_preamble.tex
LaTeX2e <2020-10-01> patch level 4
L3 programming layer <2021-02-18> (article.cls
Document Class: article 2020/04/10 v1.4m Standard LaTeX document class
(size10.clo
File: size10.clo 2020/04/10 v1.4m Standard LaTeX file (size option)
)
\c@part=\count175
\c@section=\count176
\c@subsection=\count177
\c@subsubsection=\count178
\c@paragraph=\count179
\c@subparagraph=\count180
\c@figure=\count181
\c@table=\count182
\abovecaptionskip=\skip47
\belowcaptionskip=\skip48
\bibindent=\dimen138
) (l3backend-xetex.def
File: l3backend-xetex.def 2021-03-18 L3 backend support: XeTeX
\c__kernel_sys_dvipdfmx_version_int=\count183
\l__color_backend_stack_int=\count184
\g__color_backend_stack_int=\count185
\g__graphics_track_int=\count186
\l__pdf_internal_box=\box47
\g__pdf_backend_object_int=\count187
\g__pdf_backend_annotation_int=\count188
\g__pdf_backend_link_int=\count189
)
No file _preamble.aux.
\openout1 = `_preamble.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 3.
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 3.
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 3.
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 3.
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for TS1/cmr/m/n on input line 3.
LaTeX Font Info:    Trying to load font information for TS1+cmr on input line 3
.
(ts1cmr.fd
File: ts1cmr.fd 2019/12/16 v2.5j Standard LaTeX font definitions
)
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for TU/lmr/m/n on input line 3.
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 3.
LaTeX Font Info:    ... okay on input line 3.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 3.
LaTeX Font Info:    ... okay on input line 3.
)
! Emergency stop
*** (job aborted, no legal \end found)
No pages of output.
```
</details>

**Reply by jarmos-san** [reply]
Sorry, I'm still on the  [v1 interface](https://tectonic-typesetting.github.io/book/latest/ref/v1cli.html) of `tectonic` since v2 is still under active development ([see source](https://tectonic-typesetting.github.io/book/latest/ref/v2cli.html)) & my workflow has already been setup so far. I suggest you to do the same as well until v2 of the interface is stable enough.

Here's what my project structure looks like:
```console
...
└── texfiles
    ├── build
    │   ├── main.pdf
    │   └── main.synctex.gz
    ├── subfiles
    │   ├── preface.tex
    │   ├── abstract.tex
    │   ├── section_1.tex
    │   └── section_2.tex
    └── main.tex
```

I also use the [subfiles](https://ctan.org/pkg/subfiles) package instead of the `\input` & `\include` commands because they result in unintended compilation results. And as such all the contents of the `subfiles` directory are "imported" into the `main.tex` file.

Then I invoke this command `tectonic texfiles/main.tex --outdir build --syntex` (I use VSCode, so the same command is configured accordingly with the LaTex Workshop plugin discussed above).

**Reply by ujjwal123123** [reply]
@roaldarbol This is the configuration that I tried:
```  "latex-workshop.latex.external.build.command": "tectonic",
  "latex-workshop.latex.external.build.args": ["-X", "build"],
  "latex-workshop.latex.recipes": [
    {
      "name": "tectonic",
      "tools": ["tectonic"]
    }
  ],
  "latex-workshop.latex.recipe.default": "first"
```
This configuration uses version 2, but many features including SyncTex and automatic indentation still do not work.

**Reply by jarmos-san** [reply]
Just a bit of clarification here, automatic indentation has nothing to do with `tectonic`. And afaik, [Texlabs](https://github.com/latex-lsp/texlab-vscode) has both amazing syntax highlighting & indentation support. The LaTex Workshop is more feature-rich but lacks in the aforementioned categories. :disappointed:

**Reply by MTRNord** [reply]
Hi for some reason when trying to open a preview it searches for the _preamble.pdf instead of index.pdf. Any idea how to fix this?

**Reply by jarmos-san** [reply]
@MTRNord I've not been using Tectonic for months now, so I'm not sure what went wrong. Perhaps try opening a new thread & the maintainers will help you out.

**Reply by MTRNord** [reply]
Thanks anyway :)

**Reply by albertothedev** [reply]
I am having the same issue. Were either of you able to fix it?

**Reply by MTRNord** [reply]
No I didn't spend the time to further investigate and instead just launched tectonic in watch mode and opened the PDF. Half of the times it even automatically updated.

**Reply by albertothedev** [reply]
Alright, thanks for letting me know.

**Reply by Zeta611** [reply]
@MTRNord @albertothedev It may be too late, but I found out that the cause for this issue is due to [the root file resolution mechanism of LaTeX Workshop](https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#the-root-file). I think LaTeX Workshop basically replaces the extension of the "root file"--or at least what it thinks the root file is--from `.tex` to `.pdf`. And yet LaTeX Workshop erroneously identifies the root file. This is what I believe has happened:

To find the root file, LaTeX Workshop tries to find the magic comment, then it fails (this should have happened based on the issues you described). Then it resorts to 2, finding the file with `\documentclass[...]{...}`, which happens to be `_preamble.tex` for the default configuration with Tectonic. There are a few more fallbacks, but these are not our options because the second fallback is causing the failure.

Therefore to workaround this issue, the easiest way I've found is to:
1. Set `output.name` and `output.index` to `'default'` in `Tectonic.toml`:
```toml
[[output]]
name = 'default'
index = 'default'
type = 'pdf'
```
2. Add `% !TEX root = ./default.tex` to the top of any of the files in the working directory. I chose `default.tex`. `_preamble.tex` works as well.
3. Set `"latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/default"` in `settings.json`.

I wish manually setting the root pdf path at the LaTeX Workshop extension side, but this works for me now. I'll try to file an issue or add a pull request myself to the extension when I have some time.

That's it! I hope it's not too late.

**Reply by albertothedev** [reply]
@Zeta611 

I have since switched to [Typst](https://github.com/typst/typst) :sweat_smile: but I am sure your detailed answer will help someone else.

**Reply by Zeta611** [reply]
@albertothedev Hey, from genuine curiosity, why did you switch to typst? I've been thinking to try it--is it usable to produce high quality documents?

**Reply by albertothedev** [reply]
@Zeta611 

Depends on what you mean by "high-quality documents". I suppose you are talking about document complexity, but that is a very broad spectrum even before accounting for features provided by packages so I can't give you an answer. 

Typst is currently on major version 0 so expect frequent breaking changes. It has a passionate and ambitious community, so expect huge improvements too. I see that you speak Korean and I have seen some bugs with certain languages so you might run into some yourself. I have also seen that you're plenty familiar with LaTeX. Knowing those too things it's up to you to decide whether you are willing to give it a chance or would prefer waiting for a 1.0 release.

**Reply by Zeta611** [reply]
@albertothedev Thanks a lot for the detailed answer. I really appreciate it 😀

**Reply by eljamm** [reply]
I'd like to thank everyone who has contributed to this discussion for helping me set up Tectonic in VSCode and I'd just like to add a small improvement to @Zeta611's answer.

Imagine you have two documents that use the same `_preamble` and `_postamble`.

```toml
[[output]]
name = "english"
index = "english"
type = "pdf"

[[output]]
name = "german"
index = "german"
type = "pdf"
```

Building this project results in the following structure:

```
build/
├── english
│   ├── english.pdf
│   └── ...
├── german
│   ├── german.pdf
│   └── ...
└── index
```

However, trying to open a PDF file from the LaTeX previewer results in the following message:

```
Cannot view file PDF file. File not found: file:///home/user/project/build/index/english.pdf
```

To solve this, instead of hardcoding the output directory in LaTeX Workshop, we can change it to dynamically point to the document name:

```diff
-  "latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/index",
+  "latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/%DOCFILE%",
```

This matches Tectonic's output directory using the `output.name` in the `Tectonic.toml`.

In this example, this means that the LaTeX previewer will look for the following documents:

```sh
file:///home/user/project/build/english/english.pdf
file:///home/user/project/build/german/german.pdf
```

As @Zeta611 explained, though, we need to specify `% !TEX root` for each document first or the previewer will look for `_preamble`, instead:

```latex
%% Inside src/english.tex:

% !TEX root = ./english.tex
Hello, world.
```

```latex
%% Inside src/german.tex:

% !TEX root = ./german.tex
Hallo Welt.
```

In conclusion, using `%DOCFILE%` for the `outDir` gives more naming flexibility and correct PDF previewing for both single and multi-output projects.

**Comment by LeCyberDucky** [comment]
Here's my configuration with VS Code and LaTeX Workshop. My document structure doesn't fit the default Tectonic structure, so I'm using the `compile` command for building. Also, I'm doing one initial run to get rid of warnings about undefined citations. This keeps some intermediate files around that I'm not a fan of, so I'm still trying to figure out how to get rid of those. 

```
"latex-workshop.latex.autoBuild.cleanAndRetry.enabled": false,
"latex-workshop.latex.autoBuild.run": "onSave",
"latex-workshop.latex.recipe.default": "tectonic",
"latex-workshop.latex.recipes": [
    ...
    {
        "name": "tectonic",
        "tools": [
            "tectonic-biber",
            "tectonic"
        ]
    }
],
"latex-workshop.latex.tools": [
...
    {
        "args": [
            "-X",
            "compile",
            "%DOCFILE_EXT%",
            // "-Z",
            // "shell-escape",
            "--synctex",
            "--print"
        ],
        "command": "tectonic",
        "env": {},
        "name": "tectonic"
    },
    {
        "args": [
            "-X",
            "compile",
            "%DOCFILE_EXT%",
            "-k",
            "--reruns",
            "0"//,
           // "-Z",
            // "shell-escape" // Don't include shell-escape, unless you have a very good reason for doing so
        ],
        "command": "tectonic",
        "env": {},
        "name": "tectonic-biber"
        // See: 
        // https://github.com/tectonic-typesetting/tectonic/issues/195
        // https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#latex-recipes
        // https://github.com/James-Yu/LaTeX-Workshop/wiki/FAQ#suggestions-for-bibtex-files-are-not-updated
    }
],
"latex-workshop.latex.watch.usePolling": true,
"latex-workshop.view.pdf.internal.synctex.keybinding": "double-click",
"latex-workshop.view.pdf.viewer": "tab",
```

**Reply by jarmos-san** [reply]
Thanks for sharing you configuration! I'll update my answer so that it adheres to the v2 interface of `tectonic`.

Besides, a heads up for the unwary, you might not need to pass the `-Z shell-escape` flags to `tectonic` if you're ensure of what it does. Be **ABSOLUTELY** sure you want to run arbitrary Shell commands in your LaTex files before using those flags. For more information on this, check the following resources:

- The [_"unstable options"_](https://tectonic-typesetting.github.io/book/latest/v2cli/compile.html?highlight=-Z#unstable-options) of the `tectonic -X compile` command.
- The [unofficial docs on `\write18`](https://latexref.xyz/_005cwrite18.html) to understand how you can invoke arbitrary Shell commands within LaTex files.
- An [example of how, where & why to use `\write18`](https://tex.stackexchange.com/a/20566) within your LaTex files.

**Reply by LeCyberDucky** [reply]
Thanks for the heads up. You make a good point, but in my case I do need `shell-escape`. I'll update my reponse to remove it, though, in case somebody copies it without thinking about that line.

**Reply by ratmice** [reply]
Also note that for the V2 interface IIUC shell-escape can be enabled in the [`Tectonic.toml`](https://tectonic-typesetting.github.io/book/latest/ref/tectonic-toml.html) It'd be nice to eventually hook into the trusted/untrusted workspace settings of vscode to pass this flag.

I believe `trusted` is the default unless `--untrusted` is specified, but I should probably check that.

**Comment by jarmos-san** [comment]
**NOTE**: I updated my workflow to adhere to `tectonic`'s v2 interface just a while ago & as such I'm updating the original answer with the updated configuration & and explanation. This new comment here is kept for historical & reference purpose. So, if you want to check the latest configuration, check out [this answer (comment)](https://github.com/tectonic-typesetting/tectonic/discussions/896#discussioncomment-2700291).

### The Old Answer:

Having using both [TexLab](https://github.com/latex-lsp/texlab-vscode) & [LaTex Workshop](https://github.com/James-Yu/LaTeX-Workshop), I can confidently suggest using the later even though the former's LSP suggestions & highlight is the best!

LaTex Workshop did a great job at documenting it's features, is prompt when responding to issues/queries & the plugin is **VERY** feature rich. You almost won't need the command-line at all if using the plugin. So, I suggest reading [their documentations](https://github.com/James-Yu/LaTeX-Workshop/wiki) in case you want to work with LaTex frequently on VSCode.

But regardless, I believe you need to have these properties in your `settings.json`:

```json
  "latex-workshop.latex.recipes": [
    {
      "name": "tectonic",
      "tools": ["tectonic"]
    }
  ],
  "latex-workshop.latex.tools": [
    {
      "name": "tectonic",
      "command": "tectonic",
      "args": ["%DOC%.tex", "--outdir", "%OUTDIR%", "--synctex"],
      "env": {}
    }
  ],
  "latex-workshop.latex.outDir": "build",
```
P.S: Read more about "[`latex-recipes`](https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#latex-recipes)" & "[`latex-tools`](https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#latex-tools)" before customising your VSCode-LaTex experience.

**Comment by danmackinlay** [comment]
Minor quibble: AFAIK the recommended way to write `"%DOC%.tex" is in fact "%DOC_EXT%".
